### ISSUE TITLE
An alternative style for functions and methods formatting

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rustfmt#2474]
I usually format function headers which exceed `max_width` in the following way:
```Rust
fn foo(<input args  up to max_width>)
    -> ReturnType
    where T: Foo
{
    // ..
}
```
And if function has too many input arguments:
```Rust
fn foo(<input args1 up to max_width>,
        <input args2 up to max_width>)
    -> ReturnType
    where T: Foo, // up to max_width
        T2: Bar, // up to max_width
{
    // ..
}
```

If function has a long name and/or generic parameters (e.g. >= max_width/2) "input args1" can start from the new line:

```Rust
fn log_foo<T, T2> // up to max_width/2
    (<input args1 up to max_width>,
        <input args2 up to max_width>)
    -> ReturnType
    where T: Foo, // up to max_width
        T2: Bar, // up to max_width
{
    // ..
}
```

This style visually separates input argument, output type and `where` block, while leaving the minimum amount of empty space and uses the minimum amount of lines.

AFAIK it's currently impossible to replicate such style with `.rustfmt.toml`, so I would be nice to have an option for such style. (`fn_compact_header`?)

**Comment by Centril** [comment]
Personally I'd like to style as follows sometimes (if the arrow overflows the 80 characters):
```rust
fn foo(<input args  up to max_width>)
    -> ReturnType
where
    T: Foo
{
    // ..
}
```

**Comment by nrc** [comment]
We discussed some of these options in the fmt-rfcs process. We're unlikely to support an option for newlines and indenting without a newline after the opening paren - this is just inconsistent with too much other formatting. Then the remaining issue is indeing the `where` clause. I'm not actually sure if we still support an option for this indentation, I think we might, but we did remove a lot of the options there.

I have found in the past that trying to support every possible formatting leads to a mess of options, messy implementation code, and lots of bugs for unexpected combinations. So I prefer to support fewer options nowadays. Thus something like this is unlikely to be accepted without evidence of wide support for such a thing.
